morpho-checkpoint-node/tests/Test/Morpho/QSM.hs

```diff
@@ -335,46 +354,98 @@ prop_4 = noShrinking $
           majority = 2
         }
 
-setup :: Int -> Int -> IO [NodeHandle]
-setup nodesNum testId = do
+setup :: Int -> Int -> Int -> IO [NodeHandle]
+setup nodesNum requiredMajority testId = do
   let testDir = mkTestDir testId
   removePathForcibly testDir
   createDirectoryIfMissing True testDir
-  forM [0 .. (nodesNum - 1)] $ runDualNode True testId
+  forM [0 .. (nodesNum - 1)] $ runDualNode True nodesNum requiredMajority testId
 
 cleanup :: NodeHandle -> IO ()
 cleanup NodeHandle {..} = do
   cancel mainNode
 
-runDualNode :: Bool -> Int -> Int -> IO NodeHandle
-runDualNode createDir testId nodeId = do
+runDualNode :: Bool -> Int -> Int -> Int -> Int -> IO NodeHandle
+runDualNode createDir nodesNum requiredMajority testId nodeId = do
   let testDir = mkTestDir testId
   let nodeDir = testDir ++ "/nodedir-" ++ show nodeId
-  let configDir = "tests/configuration/QSM/prop_" ++ show testId
   when createDir $ createDirectory nodeDir
-  let cliConfig =
-        (bpure (Left CliNoParser))
-          { ncTopologyFile = Right $ TopologyFile $ configDir ++ "/topology.json",
-            ncDatabaseDir = Right $ DbFile $ nodeDir ++ "/db",
-            ncNodePort = Right $ fromIntegral $ 3000 + 2 * nodeId,
-            ncValidateDb = Right True
-          }
-      configFile = configDir ++ "/config-" ++ show nodeId ++ ".yaml"
   (mockNode, rpcUpstream) <- mockRpcUpstream
 
-  nodeConfig <- getConfiguration cliConfig configFile
+  let toPublicKey =
+        fromRight' . importPublicKey . fromRight' . bytesFromHex . fromJust . normalizeHex
+  let pubKeys =
+        take
+          nodesNum
+          [ toPublicKey "ec33a3689573db2f4db4586bb7089cda045116a21cce20c9a6fe7ccadcf9fb336075b3644ac9f0a20e6d45a9e99db477cc420d050969f2d8bfb7408b2169b167",
+            toPublicKey "3ea6ebf308a16b45bbf3392d1cd226aa6858d2ea6a2402799eeffc576f0187af62ed11677c8d4d01ef03bbf638f49f2fd77e181924a52dd01b86a187af07b724"
+          ]
+
+  let toPrivKey = fromRight' . importPrivateKey . fromRight' . bytesFromHex . fromJust . normalizeHex
+  let privKey =
+        [ toPrivKey "0093e3cf4be871137e4e11b8d94ec397afb1d3b6db4cdfef01780033b7e3c67f06",
+          toPrivKey "00103becb5b909d34b5a57fc629e305c7ce8f00ef2318939e645a3c161b1d99a03"
+        ]
+          !! nodeId
+
+  systemStart <- parseTimeM True defaultTimeLocale "%FT%XZ" "2020-11-22T04:45:09Z" :: IO UTCTime
```

Comment:
For better logs I think we should always use the current time here. Then the slot indices will start from 0